### PR TITLE
feat:new discovered node record event stream

### PR DESCRIPTION
**Pull request by supernovahs**
# Solves #3701 

how should I add  [discovery_manager_tx](https://github.com/supernovahs/reth/blob/7e33ed034cc028806a6c1f45f5b66d02e2d8013b/crates/net/network/src/config.rs#L73) in **[NetworkConfigBuilder](https://github.com/supernovahs/reth/blob/7e33ed034cc028806a6c1f45f5b66d02e2d8013b/crates/net/network/src/config.rs#L125)**?   As  NetworkConfigBuilder implements Serialize which is not supported by `discovery_manager_tx: 
`

- Implement a new Struct which inherits [NetworkBuilder](https://github.com/supernovahs/reth/blob/7e33ed034cc028806a6c1f45f5b66d02e2d8013b/crates/net/network/src/config.rs#L34) struct  and discovery_manager_tx while not implementing serialize?

thx

### CODE REVIEW

**Comment by supernovahs** [comment]
cc @mattsse

File: crates/net/network/src/config.rs [hunk 1]
```diff
@@ -67,6 +69,10 @@ pub struct NetworkConfig<C> {
     pub status: Status,
     /// Sets the hello message for the p2p handshake in RLPx
     pub hello_message: HelloMessage,
+
+    pub discovery_manager_tx: UnboundedSender<DiscoveryEvent>,
+
+
```
Comment (mattsse) [inline crates/net/network/src/config.rs:75]: we don't need this here

File: crates/net/network/src/discovery.rs [hunk 2]
```diff
@@ -41,6 +45,8 @@ pub struct Discovery {
     _dns_disc_service: Option<JoinHandle<()>>,
     /// Events buffered until polled.
     queued_events: VecDeque<DiscoveryEvent>,
+
+    discovery_manager_tx: UnboundedSender<DiscoveryEvent>,
```
Comment (mattsse) [inline crates/net/network/src/discovery.rs:49]: there can be multiple listeners
what we need is a list of senders just like this:

https://github.com/paradigmxyz/reth/blob/b07dfb55259db022455e267900a02e599869b78b/crates/net/discv4/src/lib.rs#L392-L393

and 

https://github.com/paradigmxyz/reth/blob/b07dfb55259db022455e267900a02e599869b78b/crates/net/discv4/src/lib.rs#L686-L695

**Review by mattsse — CHANGES_REQUESTED** [review]
thanks for taking this on,

I'd like to take a slightly different approach than extracting the discovered event variant and instead add a list of listeners that are updated when this variant is enqueued

similar to how this is already implemented in the Discv4 impl

so the when the networkhandle sends a message to create a new listener, the Discovery struct creates a new channel, keeps the sender and returns the receiver

does that make sense?

**Comment by supernovahs** [comment]
I have made some changes as you mentioned.

File: crates/net/network/src/discovery.rs [hunk 3]
```diff
@@ -96,6 +103,23 @@ impl Discovery {
         })
     }
 
+    pub fn register_listener(&mut self) -> mpsc::Receiver<DiscoveryEvent> {
```
Comment (mattsse) [inline crates/net/network/src/discovery.rs:106]: now we need to call this from the NetworkManager and add a new networkhandlemessage variant

just like we do for this

https://github.com/paradigmxyz/reth/blob/dbafe23cce45917a43f6640d71a0216fe3268428/crates/net/network/src/network.rs#L292-L293

**Review by mattsse — CHANGES_REQUESTED** [review]
great progress

**Comment by supernovahs** [comment]
I have made the necessary changes

File: crates/net/network/src/discovery.rs [hunk 4]
```diff
@@ -246,7 +248,7 @@ impl Discovery {
             _dns_discovery: None,
             dns_discovery_updates: None,
             _dns_disc_service: None,
-             discovery_listeners: Vec::with_capacity(1),
+            discovery_listeners: Vec::default()
```
Comment (mattsse) [inline crates/net/network/src/discovery.rs:251]:
```suggestion
            discovery_listeners: Default::default()
```

File: crates/net/network/src/discovery.rs [hunk 5]
```diff
@@ -90,7 +90,7 @@ impl Discovery {
             };
 
         Ok(Self {
-            discovery_listeners: Vec::with_capacity(1),  
+            discovery_listeners: Vec::default(),
```
Comment (mattsse) [inline crates/net/network/src/discovery.rs:93]: pedantic style nit:

```suggestion
            discovery_listeners: Default::default(),  
```

File: crates/primitives/src/listener/discovery_listeners.rs [hunk 6]
```diff
@@ -0,0 +1,36 @@
+use tokio::sync::mpsc;
+use tokio_stream::wrappers::UnboundedReceiverStream;
+
+/// A collection of event listeners for a task.
+#[derive(Clone, Debug)]
+pub struct DiscoveryListeners<T> {
+    /// All listeners for events
+    listeners: Vec<mpsc::UnboundedSender<T>>,
```
Comment (mattsse) [inline crates/primitives/src/listener/discovery_listeners.rs:8]: I'm okay with making this unbounded

File: crates/primitives/src/listener/discovery_listeners.rs
Comment (mattsse) [inline crates/primitives/src/listener/discovery_listeners.rs:1]: this does not belong in primitives.

I'd prefer the previous impl which just had a simple Vec<Sender> field and a notify function.
It's okay duplicating these things imo

**Review by mattsse — CHANGES_REQUESTED** [review]
nice!

a few nits,

**Comment by supernovahs** [comment]
I have made the changes. I hope you meant to change the position of disoovery_listeners in the last comment you made.
   thx

File: crates/net/network/src/manager.rs [hunk 7]
```diff
@@ -919,3 +928,12 @@ pub enum NetworkEvent {
...
     PeerRemoved(PeerId),
 }
+
+#[derive(Debug, Clone)]
+pub enum DiscoveredEvent{
+    EventQueued{
+        peer_id: PeerId,
+        socket_addr: SocketAddr,
+        fork_id: Option<ForkId>
+    }
```
Comment (mattsse) [inline crates/net/network/src/manager.rs:938]: we don't need this

File: crates/net/network/src/discovery.rs [hunk 8]
```diff
@@ -219,14 +248,18 @@ impl Discovery {
...
     }
 }
 
 /// Events produced by the [`Discovery`] manager.
+ #[derive(Debug, Clone)]
 pub enum DiscoveryEvent {
-    /// A new node was discovered
-    Discovered { peer_id: PeerId, socket_addr: SocketAddr, fork_id: Option<ForkId> },
+
+    NewNode(DiscoveredEvent),
```
Comment (mattsse) [inline crates/net/network/src/discovery.rs:260]: why are we using a new variant here?

we don't need this internally

File: crates/net/network/src/discovery_listeners.rs
Comment (mattsse) [inline crates/net/network/src/discovery_listeners.rs:1]: this is now redundant and unused?

can be removed

**Review by mattsse — CHANGES_REQUESTED** [review]
almost there, just needs cleanup

remove new discovery event variant, that we don't need

File: crates/net/network/src/manager.rs (hunk 7 above)
Comment (supernovahs) [reply crates/net/network/src/manager.rs:938]: If we remove this , then how will we encapsulate `peer_id`, `socket_addr`, and `fork_id`?

Comment (mattsse) [reply crates/net/network/src/manager.rs:938]: oh I get it now, very good point!

will give it another look, I think this is good now, pending the listner cleanup,

thanks

**Review by mattsse — APPROVED** [review]
tysm for this!

I made a simplification re how the listeners are stored

**Comment by supernovahs** [comment]
Thx for helping always

**Pull request merged**